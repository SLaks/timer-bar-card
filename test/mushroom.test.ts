import { multiply, HomeAssistant, PlaywrightBrowser, PlaywrightElement } from "hass-taste-test";
import { synchronizeTimerPaused, toMatchDualSnapshot } from "./util";
expect.extend({ toMatchDualSnapshot });

const CONFIGURATION_YAML = `
timer:
  ${multiply(10, (i) => `
  test${i}:
    duration: "00:${i}:00"
`)}
`;

let hass: HomeAssistant<PlaywrightElement>;

beforeAll(async () => {
  hass = await HomeAssistant.create(CONFIGURATION_YAML, {
    browser: new PlaywrightBrowser(process.env.BROWSER || "firefox"),
  });
  await hass.addResource(__dirname + "/../dist/timer-bar-card.js", "module");
  await hass.addResource(__dirname + '/resources/mushroom.js', 'module');
}, 30000);
afterAll(async () => await hass.close());

it("Basic mushroom", async () => {
  const dashboard = await hass.Dashboard([
    { type: "custom:timer-bar-card", entity: "timer.test1", mushroom: true },
  ]);
  const card = dashboard.cards[0];
  await expect(card).toMatchDualSnapshot("idle");

  await hass.callService("timer", "start", {}, { entity_id: "timer.test1" });
  await expect(card).toMatchDualSnapshot("running");

  await synchronizeTimerPaused(hass, "timer.test1", "00:00:58");
  await hass.callService("timer", "pause", {}, { entity_id: "timer.test1" });
  await expect(card).toMatchDualSnapshot("paused");
});

it("Vertical mushroom", async () => {
  const dashboard = await hass.Dashboard([
    { type: "custom:timer-bar-card", entity: "timer.test2", mushroom: { layout: 'vertical' } },
  ]);
  const card = dashboard.cards[0];
  await expect(card).toMatchDualSnapshot("idle");

  await hass.callService("timer", "start", {}, { entity_id: "timer.test2" });
  await expect(card).toMatchDualSnapshot("running");

  await synchronizeTimerPaused(hass, "timer.test2", "00:01:58");
  await hass.callService("timer", "pause", {}, { entity_id: "timer.test2" });
  await expect(card).toMatchDualSnapshot("paused");
});

it("Mushroom customization", async () => {
  const dashboard = await hass.Dashboard([{
    type: "grid",
    columns: 2,
    square: false,
    cards: [{
      type: "custom:timer-bar-card",
      entity: "timer.test3",
      mushroom: {
        primary_info: 'state',
        secondary_info: 'none'
      }
    }, {
      type: "custom:timer-bar-card",
      entity: "timer.test3",
      name: 'No icon',
      mushroom: {
        icon_type: 'none',
      }
    }],
  }]);
  const card = dashboard.cards[0];
  await hass.callService("timer", "start", {}, { entity_id: "timer.test3" });
  await hass.callService("timer", "finish", {}, { entity_id: "timer.test3" });
  await expect(card).toMatchDualSnapshot("infos");
});

it("Mushroom fill_container", async () => {
  const dashboard = await hass.Dashboard([{
    type: "grid",
    columns: 2,
    square: true,
    cards: [{
      type: "custom:timer-bar-card",
      entity: "timer.test4",
      mushroom: {
        fill_container: false
      }
    }, {
      type: "custom:timer-bar-card",
      entity: "timer.test4",
      mushroom: {
        fill_container: true
      }
    }],
  }]);
  const card = dashboard.cards[0];
  await expect(card).toMatchDualSnapshot("fill_container");
});
