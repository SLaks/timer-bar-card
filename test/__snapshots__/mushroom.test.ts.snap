// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic mushroom: idle 1`] = `
"<timer-bar-card hass="[object Object]">
  <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="true">
    <ha-card raised="false">
      <slot></slot>
      <mushroom-card appearance="[object Object]">
        <div class="container">
          <slot></slot>
        </div>
        <mushroom-state-item appearance="[object Object]">
          <div class="container">
            <div class="icon">
              <slot name="icon"></slot>
              <slot name="badge"></slot>
            </div>
            <div class="info">
              <slot name="info"></slot>
            </div>
          </div>
          <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
            <div class="shape">
              <ha-icon icon="mdi:timer-outline">…</ha-icon>
            </div>
          </mushroom-shape-icon>
          <div class="container" slot="info">
            <span class="primary">
              test1
            </span>
            <span class="default secondary">
              <div class="pointer text-content value">
                idle
              </div>
            </span>
          </div>
        </mushroom-state-item>
      </mushroom-card>
    </ha-card>
  </timer-bar-mushroom-row>
</timer-bar-card>
"
`;

exports[`Basic mushroom: paused 1`] = `
"<timer-bar-card hass="[object Object]">
  <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="true">
    <ha-card raised="false">
      <slot></slot>
      <mushroom-card appearance="[object Object]">
        <div class="container">
          <slot></slot>
        </div>
        <mushroom-state-item appearance="[object Object]">
          <div class="container">
            <div class="icon">
              <slot name="icon"></slot>
              <slot name="badge"></slot>
            </div>
            <div class="info">
              <slot name="info"></slot>
            </div>
          </div>
          <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
            <div class="shape">
              <ha-icon icon="mdi:timer-outline">…</ha-icon>
            </div>
          </mushroom-shape-icon>
          <div class="container" slot="info">
            <span class="primary">
              test1
            </span>
            <span class="default secondary">
              <div class="pointer status">
                paused
              </div>
              <div class="pointer text-content value">
                58
              </div>
            </span>
          </div>
        </mushroom-state-item>
      </mushroom-card>
    </ha-card>
  </timer-bar-mushroom-row>
</timer-bar-card>
"
`;

exports[`Basic mushroom: running 1`] = `
"<timer-bar-card hass="[object Object]">
  <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="true">
    <ha-card raised="false">
      <slot></slot>
      <mushroom-card appearance="[object Object]">
        <div class="container">
          <slot></slot>
        </div>
        <mushroom-state-item appearance="[object Object]">
          <div class="container">
            <div class="icon">
              <slot name="icon"></slot>
              <slot name="badge"></slot>
            </div>
            <div class="info">
              <slot name="info"></slot>
            </div>
          </div>
          <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
            <div class="shape">
              <ha-icon icon="mdi:timer-outline">…</ha-icon>
            </div>
          </mushroom-shape-icon>
          <div class="container" slot="info">
            <span class="primary">
              test1
            </span>
            <span class="default secondary">
              <div class="bar-container pointer">
                <div class="bar">
                  <div></div>
                </div>
              </div>
              <div class="pointer text-content value">
                59
              </div>
            </span>
          </div>
        </mushroom-state-item>
      </mushroom-card>
    </ha-card>
  </timer-bar-mushroom-row>
</timer-bar-card>
"
`;

exports[`Mushroom customization: infos 1`] = `
"<hui-grid-card _cards="[object HTMLElement],[object HTMLElement]" editMode="false">
  <div id="root">
    <timer-bar-card hass="[object Object]">
      <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="[object Object]">
        <ha-card raised="false">
          <slot></slot>
          <mushroom-card appearance="[object Object]">
            <div class="container">
              <slot></slot>
            </div>
            <mushroom-state-item appearance="[object Object]">
              <div class="container">
                <div class="icon">
                  <slot name="icon"></slot>
                  <slot name="badge"></slot>
                </div>
                <div class="info">
                  <slot name="info"></slot>
                </div>
              </div>
              <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
                <div class="shape">
                  <ha-icon icon="mdi:timer-outline">…</ha-icon>
                </div>
              </mushroom-shape-icon>
              <div class="container" slot="info">
                <span class="primary">
                  idle
                </span>
                <span class="default secondary">
                  <div class="pointer text-content value"></div>
                </span>
              </div>
            </mushroom-state-item>
          </mushroom-card>
        </ha-card>
      </timer-bar-mushroom-row>
    </timer-bar-card>
    <timer-bar-card hass="[object Object]">
      <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="[object Object]">
        <ha-card raised="false">
          <slot></slot>
          <mushroom-card appearance="[object Object]">
            <div class="container no-icon">
              <slot></slot>
            </div>
            <mushroom-state-item appearance="[object Object]">
              <div class="container">
                <div class="info">
                  <slot name="info"></slot>
                </div>
              </div>
              <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
                <div class="shape">
                  <ha-icon icon="mdi:timer-outline">…</ha-icon>
                </div>
              </mushroom-shape-icon>
              <div class="container" slot="info">
                <span class="primary">
                  No icon
                </span>
                <span class="default secondary">
                  <div class="pointer text-content value">
                    idle
                  </div>
                </span>
              </div>
            </mushroom-state-item>
          </mushroom-card>
        </ha-card>
      </timer-bar-mushroom-row>
    </timer-bar-card>
  </div>
</hui-grid-card>
"
`;

exports[`Vertical mushroom: idle 1`] = `
"<timer-bar-card hass="[object Object]">
  <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="[object Object]">
    <ha-card raised="false">
      <slot></slot>
      <mushroom-card appearance="[object Object]">
        <div class="container">
          <slot></slot>
        </div>
        <mushroom-state-item appearance="[object Object]">
          <div class="container vertical">
            <div class="icon">
              <slot name="icon"></slot>
              <slot name="badge"></slot>
            </div>
            <div class="info">
              <slot name="info"></slot>
            </div>
          </div>
          <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
            <div class="shape">
              <ha-icon icon="mdi:timer-outline">…</ha-icon>
            </div>
          </mushroom-shape-icon>
          <div class="container" slot="info">
            <span class="primary">
              test2
            </span>
            <span class="secondary vertical">
              <div class="pointer text-content value">
                idle
              </div>
            </span>
          </div>
        </mushroom-state-item>
      </mushroom-card>
    </ha-card>
  </timer-bar-mushroom-row>
</timer-bar-card>
"
`;

exports[`Vertical mushroom: paused 1`] = `
"<timer-bar-card hass="[object Object]">
  <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="[object Object]">
    <ha-card raised="false">
      <slot></slot>
      <mushroom-card appearance="[object Object]">
        <div class="container">
          <slot></slot>
        </div>
        <mushroom-state-item appearance="[object Object]">
          <div class="container vertical">
            <div class="icon">
              <slot name="icon"></slot>
              <slot name="badge"></slot>
            </div>
            <div class="info">
              <slot name="info"></slot>
            </div>
          </div>
          <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
            <div class="shape">
              <ha-icon icon="mdi:timer-outline">…</ha-icon>
            </div>
          </mushroom-shape-icon>
          <div class="container" slot="info">
            <span class="primary">
              test2
            </span>
            <span class="secondary vertical">
              <div class="pointer status">
                paused
              </div>
              <div class="pointer text-content value">
                1:58
              </div>
            </span>
          </div>
        </mushroom-state-item>
      </mushroom-card>
    </ha-card>
  </timer-bar-mushroom-row>
</timer-bar-card>
"
`;

exports[`Vertical mushroom: running 1`] = `
"<timer-bar-card hass="[object Object]">
  <timer-bar-mushroom-row config="[object Object]" hass="[object Object]" mushroom="[object Object]">
    <ha-card raised="false">
      <slot></slot>
      <mushroom-card appearance="[object Object]">
        <div class="container">
          <slot></slot>
        </div>
        <mushroom-state-item appearance="[object Object]">
          <div class="container vertical">
            <div class="icon">
              <slot name="icon"></slot>
              <slot name="badge"></slot>
            </div>
            <div class="info">
              <slot name="info"></slot>
            </div>
          </div>
          <mushroom-shape-icon disabled="false" icon="mdi:timer-outline" slot="icon">
            <div class="shape">
              <ha-icon icon="mdi:timer-outline">…</ha-icon>
            </div>
          </mushroom-shape-icon>
          <div class="container" slot="info">
            <span class="primary">
              test2
            </span>
            <span class="secondary vertical">
              <div class="bar-container pointer">
                <div class="bar">
                  <div></div>
                </div>
              </div>
              <div class="pointer text-content value">
                1:59
              </div>
            </span>
          </div>
        </mushroom-state-item>
      </mushroom-card>
    </ha-card>
  </timer-bar-mushroom-row>
</timer-bar-card>
"
`;
